File: ApolloCore/src/main/java/me/solarbam/apollo/apolloCore/utils/BlockUtils.java
package me.solarbam.apollo.apolloCore.utils;

import org.bukkit.Location;
import org.bukkit.Material;
import java.util.ArrayList;
import java.util.List;

public final class BlockUtils {

    private BlockUtils() {}

    public static Pair<Location, Location> minMax(Location primary, Location secondary) {
        double minX = Math.min(primary.getX(), secondary.getX());
        double minY = Math.min(primary.getY(), secondary.getY());
        double minZ = Math.min(primary.getZ(), secondary.getZ());
        double maxX = Math.max(primary.getX(), secondary.getX());
        double maxY = Math.max(primary.getY(), secondary.getY());
        double maxZ = Math.max(primary.getZ(), secondary.getZ());
        return new Pair<>(
            new Location(primary.getWorld(), minX, minY, minZ),
            new Location(primary.getWorld(), maxX, maxY, maxZ)
        );
    }

    public static List<Location> boundingBox(Location primary, Location secondary) {
        List<Location> locations = new ArrayList<>();
        Pair<Location, Location> pair = minMax(primary, secondary);
        for (int x = (int) pair.first.getX(); x <= (int) pair.second.getX(); x++) {
            for (int y = (int) pair.first.getY(); y <= (int) pair.second.getY(); y++) {
                for (int z = (int) pair.first.getZ(); z <= (int) pair.second.getZ(); z++) {
                    locations.add(new Location(primary.getWorld(), x, y, z));
                }
            }
        }
        return locations;
    }

    public static List<ApolloVector> plotLine(Location alpha, Location bravo) {
        return plotLine(ApolloVector.fromLocation(alpha), ApolloVector.fromLocation(bravo));
    }

    public static List<ApolloVector> plotLine(org.bukkit.util.Vector alpha, org.bukkit.util.Vector bravo) {
        return plotLine(ApolloVector.fromBukkit(alpha), ApolloVector.fromBukkit(bravo));
    }

    public static List<ApolloVector> plotLine(ApolloVector alpha, ApolloVector bravo) {
        List<ApolloVector> locations = new ArrayList<>();
        ApolloVector delta = bravo.subtract(alpha);
        double length = delta.magnitude();
        ApolloVector direction = delta.divide(new ApolloVector(length, length, length));
        for (int i = 0; i < (int) length; i++) {
            ApolloVector location = alpha.add(direction.multiply(new ApolloVector(i, i, i)));
            locations.add(location);
        }
        return locations;
    }

    public static Location getMinLocation(Location primary, Location secondary) {
        double minX = Math.min(primary.getX(), secondary.getX());
        double minY = Math.min(primary.getY(), secondary.getY());
        double minZ = Math.min(primary.getZ(), secondary.getZ());
        return new Location(primary.getWorld(), minX, minY, minZ);
    }

    public static Location getMaxLocation(Location primary, Location secondary) {
        double maxX = Math.max(primary.getX(), secondary.getX());
        double maxY = Math.max(primary.getY(), secondary.getY());
        double maxZ = Math.max(primary.getZ(), secondary.getZ());
        return new Location(primary.getWorld(), maxX, maxY, maxZ);
    }

    public static Location getCenterLocation(Location primary, Location secondary) {
        Pair<Location, Location> pair = minMax(primary, secondary);
        Location minPair = pair.first;
        Location maxPair = pair.second;
        double x = minPair.getX() + (maxPair.getX() - minPair.getX()) / 2;
        double y = minPair.getY() + (maxPair.getY() - minPair.getY()) / 2;
        double z = minPair.getZ() + (maxPair.getZ() - minPair.getZ()) / 2;
        return new Location(primary.getWorld(), x, y, z);
    }

    public static int getHeight(Location primary, Location secondary) {
        return (int) (Math.max(primary.getY(), secondary.getY()) - Math.min(primary.getY(), secondary.getY()));
    }

    public static int getWidth(Location primary, Location secondary) {
        return (int) (Math.max(primary.getX(), secondary.getX()) - Math.min(primary.getX(), secondary.getX()));
    }

    public static int getLength(Location primary, Location secondary) {
        return (int) (Math.max(primary.getZ(), secondary.getZ()) - Math.min(primary.getZ(), secondary.getZ()));
    }

    public static int getVolume(Location primary, Location secondary) {
        return getHeight(primary, secondary) * getWidth(primary, secondary) * getLength(primary, secondary);
    }

    public static int getSurfaceArea(Location primary, Location secondary) {
        int height = getHeight(primary, secondary);
        int width = getWidth(primary, secondary);
        int length = getLength(primary, secondary);
        return 2 * (height * width + height * length + width * length);
    }

    public static double getDiagonal(Location primary, Location secondary) {
        return primary.distance(secondary);
    }

    public static boolean equalsLocation(Location primary, Location secondary) {
        return primary.getX() == secondary.getX() &&
               primary.getY() == secondary.getY() &&
               primary.getZ() == secondary.getZ();
    }

    public static boolean isSafeSpawnLocation(Location location) {
        return isSafeSpawnLocation(location.clone(), 2);
    }

    public static boolean isSafeSpawnLocation(Location location, int height) {
        for (int i = 0; i < height; i++) {
            Location check = location.clone().add(0.0, i, 0.0);
            if (check.getBlock().getType() != Material.AIR) {
                return false;
            }
        }
        return true;
    }

    public static void shootBlock(Location location, ApolloVector velocity, ApolloVector direction) {
        var fallingBlock = WorldUtils.spawnFallingBlock(location, Material.STONE);
        fallingBlock.setVelocity(ApolloVector.toBukkit(velocity.multiply(direction)));
    }

    // Helper Pair class
    public static class Pair<F, S> {
        public final F first;
        public final S second;
        public Pair(F first, S second) {
            this.first = first;
            this.second = second;
        }
    }
}

File: ApolloCore/src/main/kotlin/me/solar/apollo/apolloBukkitCore/utils/BlockUtils.kt
@file:JvmName("BlockUtils")
package me.solar.apollo.apolloBukkitCore.utils

import org.bukkit.Location
import org.bukkit.Material
import kotlin.math.max
import kotlin.math.min

fun minMax(primary: Location, secondary: Location): Pair<Location, Location> {
    val minX = min(primary.x, secondary.x)
    val minY = min(primary.y, secondary.y)
    val minZ = min(primary.z, secondary.z)

    val maxX = max(primary.x, secondary.x)
    val maxY = max(primary.y, secondary.y)
    val maxZ = max(primary.z, secondary.z)

    return Pair(
        Location(primary.world, minX, minY, minZ),
        Location(primary.world, maxX, maxY, maxZ)
    )
}

fun boundingBox(primary: Location, secondary: Location): List<Location> {
    val locations = mutableListOf<Location>()

    val pair = minMax(primary, secondary)

    for (x in pair.first.x.toInt()..pair.second.x.toInt()) {
        for (y in pair.first.y.toInt()..pair.second.y.toInt()) {
            for (z in pair.first.z.toInt()..pair.second.z.toInt()) {
                locations.add(Location(primary.world, x.toDouble(), y.toDouble(), z.toDouble()))
            }
        }
    }
    return locations
}

fun Location.isInBoundingBox(primary: Location, secondary: Location): Boolean {
    val pair = minMax(primary, secondary)
    return x in pair.first.x..pair.second.x && y >= pair.first.y
            && y <= pair.second.y && z >= pair.first.z && z <= pair.second.z
}

fun plotLine(alpha: Location, bravo: Location): List<ApolloVector> {
    return plotLine(alpha.toApolloVector(), bravo.toApolloVector())
}

fun plotLine(alpha: org.bukkit.util.Vector, bravo: org.bukkit.util.Vector): List<ApolloVector> {
    return plotLine(alpha.toApolloVector(), bravo.toApolloVector())
}

fun plotLine(alpha: ApolloVector, bravo: ApolloVector): List<ApolloVector> {
    val locations = mutableListOf<ApolloVector>()
    val delta = bravo.subtract(alpha)
    val length = delta.magnitude()
    val direction = delta.divide(ApolloVector(length, length, length))

    for (i in 0 until length.toInt()) {
        val location = alpha.add(direction.multiply(ApolloVector(i.toDouble(), i.toDouble(), i.toDouble())))
        locations.add(location)
    }
    return locations
}

// --- Location Methods ---
fun getMinLocation(primary: Location, secondary: Location): Location {
    val minX = min(primary.x, secondary.x)
    val minY = min(primary.y, secondary.y)
    val minZ = min(primary.z, secondary.z)
    return Location(primary.world, minX, minY, minZ)
}

fun getMaxLocation(primary: Location, secondary: Location): Location {
    val maxX = max(primary.x, secondary.x)
    val maxY = max(primary.y, secondary.y)
    val maxZ = max(primary.z, secondary.z)
    return Location(primary.world, maxX, maxY, maxZ)
}

fun getCenterLocation(primary: Location, secondary: Location): Location {
    val pair = minMax(primary, secondary)
    val minPair = pair.first
    val maxPair = pair.second

    val x = minPair.x + (maxPair.x - minPair.x) / 2
    val y = minPair.y + (maxPair.y - minPair.y) / 2
    val z = minPair.z + (maxPair.z - minPair.z) / 2

    return Location(primary.world, x, y, z)
}

fun getHeight(primary: Location, secondary: Location): Int {
    return (max(primary.y, secondary.y) - min(primary.y, secondary.y)).toInt()
}

fun getWidth(primary: Location, secondary: Location): Int {
    return (max(primary.x, secondary.x) - min(primary.x, secondary.x)).toInt()
}

fun getLength(primary: Location, secondary: Location): Int {
    return (max(primary.z, secondary.z) - min(primary.z, secondary.z)).toInt()
}

fun getVolume(primary: Location, secondary: Location): Int {
    return getHeight(primary, secondary) * getWidth(primary, secondary) * getLength(primary, secondary)
}

fun getSurfaceArea(primary: Location, secondary: Location): Int {
    return 2 * (getHeight(primary, secondary) * getWidth(primary, secondary) + getHeight(
        primary,
        secondary
    ) * getLength(primary, secondary) + getWidth(primary, secondary) * getLength(primary, secondary))
}

fun getDiagonal(primary: Location, secondary: Location): Double {
    return primary.distance(secondary)
}

fun equalsLocation(primary: Location, secondary: Location): Boolean {
    return primary.x == secondary.x && primary.y == secondary.y && primary.z == secondary.z
}

fun isSafeSpawnLocation(location: Location): Boolean {
    return isSafeSpawnLocation(location.clone(), 2)
}

fun isSafeSpawnLocation(location: Location, height: Int): Boolean {
    for (i in 0 until height) {
        if (location.add(0.0, i.toDouble(), 0.0).block.type != Material.AIR) {
            return false
        }
    }
    return true
}

fun shootBlock(location: Location, velocity: ApolloVector, direction: ApolloVector) {
    val fallingBlock = WorldUtils.spawnFallingBlock(location, Material.STONE)
    fallingBlock.velocity = (velocity.multiply(direction)).toBukkit()
}